R/obic.R

```diff
@@ -574,66 +600,235 @@ obic_field_dt <- function(dt,output = 'all') {
+  th_obi_p = sort(th_obi_p[th_obi_p > 0 & th_obi_p <= 1])
+  th_obi_b = sort(th_obi_b[th_obi_b > 0 & th_obi_b <= 1])
+  th_obi_e = sort(th_obi_e[th_obi_e > 0 & th_obi_e <= 1])
+  th_obi_m = sort(th_obi_m[th_obi_m > 0 & th_obi_m <= 1])
+  
+  # the number of threshold classes required
+  nclass <- max(length(th_obi_c),length(th_obi_p),length(th_obi_b),length(th_obi_e),length(th_obi_m))
+  
+  # add checkmate for thresholds
+  checkmate::assert_numeric(th_obi_c, lower = 0, upper = 1, any.missing = FALSE, len = nclass)
+  checkmate::assert_numeric(th_obi_p, lower = 0, upper = 1, any.missing = FALSE, len = nclass)
+  checkmate::assert_numeric(th_obi_b, lower = 0, upper = 1, any.missing = FALSE, len = nclass)
+  checkmate::assert_numeric(th_obi_e, lower = 0, upper = 1, any.missing = FALSE, len = nclass)
+  checkmate::assert_numeric(th_obi_m, lower = 0, upper = 1, any.missing = FALSE, len = nclass)
+  
+  # calculate obic score for all the fields
+  out <- obic_field_dt(dt = dt, output = c('scores','indicators'))
+  
+  # aggregate into a farm for indicators and scores, and melt
+  dt.farm <- copy(out)
+  dt.farm[, farmid := 1]
+  cols <- colnames(dt.farm)[grepl('^RM_',colnames(dt.farm))]
+  if(length(cols)>0){dt.farm[,c(cols):=NULL]}
+  dt.farm <- dt.farm[,lapply(.SD,as.numeric)]
+  dt.farm <- melt(dt.farm,
+                  id.vars = c('farmid','ID'), 
+                  variable.name = 'indicator',
+                  value.name = 'obi_score')
+  
+  # add threshold columns
+  nclass <- paste0('nclass_',1:nclass)
+  
+  # add thresholds
+  dt.farm[grepl('^I_C|^S_C',indicator),c(nclass) := as.list(th_obi_c)]
+  dt.farm[grepl('^I_P|^S_P|^I_BCS',indicator),c(nclass) := as.list(th_obi_p)]
+  dt.farm[grepl('^I_B|^S_B',indicator) & !grepl('^I_BCS',indicator),c(nclass) := as.list(th_obi_b)]
+  dt.farm[grepl('^I_M|^S_M',indicator),c(nclass) := as.list(th_obi_m)]
+  dt.farm[grepl('^I_E|^S_E',indicator),c(nclass) := as.list(th_obi_e)]
+  dt.farm[grepl('^S_T_OBI',indicator),c(nclass) := as.list((th_obi_c + th_obi_p + th_obi_b)/3)]
+  
+  # melt the threshold values in farm data.table
+  dt.farm2 <- melt(dt.farm, 
+                   id.vars = c('farmid','ID','indicator','obi_score'), 
+                   variable.name = 'threshold')
+  
+  # when score or indicator is NA, then not applicable, so then distance to target is zero (and score equal to one)
+  dt.farm2[is.na(obi_score), obi_score := 1]
+  
+  # count number of fields per indicator and score
+  dt.farm2[, catvalue := fifelse(obi_score <= value, 1, 0)]
+  dt.farm2[, catvalue := cumsum(catvalue),by = c('ID','indicator')]
+  dt.farm2[catvalue > 1, catvalue := 0]
+  
+  # estimate the number of fields in a given class
+  dt.farm2 <- dcast(dt.farm2, 
+                    indicator ~ threshold, 
+                    value.var = 'catvalue',
+                    fun.aggregate = sum, na.rm=T)
+  
+  # change names
+  setnames(dt.farm2,gsub('nclass_','s_obi_farm_',colnames(dt.farm2)))
```

Comment:
These names can be confusing. At first I thought it was about multiple farms. If classes are not set dynamically these numbers could be replaced by class names